test/helpers/kubectl.go

```diff
@@ -666,11 +666,17 @@ func (kub *Kubectl) CiliumReport(namespace string, pod string, commands []string
 // CheckLogsForDeadlock checks if the logs for Cilium log messages that signify
 // that a deadlock has occurred.
 func (kub *Kubectl) CheckLogsForDeadlock() {
-	deadlockCheckCmd := fmt.Sprintf("%s -n %s logs -l k8s-app=cilium | grep -qi -B 5 -A 5 deadlock", KubectlCmd, KubeSystemNamespace)
+	deadlockCheckCmd := fmt.Sprintf("%s -n %s --timestamps=true logs -l k8s-app=cilium | grep -qi -B 5 -A 5 deadlock", KubectlCmd, KubeSystemNamespace)
 	res := kub.Exec(deadlockCheckCmd)
 	if res.WasSuccessful() {
 		log.Errorf("Deadlock during test run detected, check Cilium logs for context")
 	}
+	// Also check for previous container
+	deadlockCheckCmd = fmt.Sprintf("%s -n %s --timestamps=true --previous logs -l k8s-app=cilium | grep -qi -B 5 -A 5 deadlock", KubectlCmd, KubeSystemNamespace)
```

Comment:
This is not going to work no?

```
vagrant@k8s1:~$ kubectl -n kube-system logs --timestamps=true --previous -l k8s-app=cilium                   
Error from server (BadRequest): previous terminated container "cilium-agent" in pod "cilium-q9zmz" not found 
vagrant@k8s1:~$ 
```

Reply:
@eloycoto was there a previously terminated container ? If there is no previous container, that log might be emitted.

Reply:
```
vagrant@k8s1:~$ kubectl -n kube-system logs --timestamps=true --previous -l k8s-app=cilium
Error from server (BadRequest): previous terminated container "cilium-agent" in pod "cilium-xk6fh" not found
vagrant@k8s1:~$ kubectl get pods --all-namespaces
NAMESPACE     NAME                           READY     STATUS    RESTARTS   AGE
default       netcatds-slz6s                 1/1       Running   0          20h
kube-system   cilium-xk6fh                   1/1       Running   0          20h
kube-system   etcd-k8s1                      1/1       Running   0          22h
kube-system   kube-apiserver-k8s1            1/1       Running   0          22h
kube-system   kube-controller-manager-k8s1   1/1       Running   1          22h
kube-system   kube-dns-56b7b56cc6-vvz9d      3/3       Running   0          22h
kube-system   kube-proxy-tspmn               1/1       Running   0          22h
kube-system   kube-scheduler-k8s1            1/1       Running   0          22h
vagrant@k8s1:~$ kubectl delete -n kube-system -l k8s-app=cilium
error: at least one resource must be specified to use a selector
vagrant@k8s1:~$ kubectl delete -n kube-system pods -l k8s-app=cilium
pod "cilium-xk6fh" deleted
vagrant@k8s1:~$ kubectl -n kube-system logs --timestamps=true --previous -l k8s-app=cilium
Error from server (BadRequest): previous terminated container "cilium-agent" in pod "cilium-z28jd" not found
vagrant@k8s1:~$ kubectl get pods --all-namespaces
NAMESPACE     NAME                           READY     STATUS    RESTARTS   AGE
default       netcatds-slz6s                 1/1       Running   0          20h
kube-system   cilium-z28jd                   1/1       Running   0          12s
kube-system   etcd-k8s1                      1/1       Running   0          22h
kube-system   kube-apiserver-k8s1            1/1       Running   0          22h
kube-system   kube-controller-manager-k8s1   1/1       Running   1          22h
kube-system   kube-dns-56b7b56cc6-vvz9d      3/3       Running   0          22h
kube-system   kube-proxy-tspmn               1/1       Running   0          22h
kube-system   kube-scheduler-k8s1            1/1       Running   0          22h
```

If I delete a container I didn't see the error message too :-(

Reply:
I'm unsure whether that is valid, as I'm unsure of the semantics of `--previous` - might deletion also delete the previous pod's logs?  I.e., using `--previous` only works if the pod fails its healthcheck / the application inside has crashed?

Reply:
From the Kubernetes CLI documentation: 
" -p, --previous[=false]: If true, print the logs for the previous instance of the container in a pod if it exists."

Link: https://kubernetes-v1-4.github.io/docs/user-guide/kubectl/kubectl_logs/

I don't see how this wouldn't work with the fix proposed here.

Reply:
The main issue @ianvernon is that if we use `previous` and fails we do not get any logs and the same way that I deleted that pods are how we test in Chaos test.

Reply:
'we do not get any logs' what logs are you referring to? The tests don't error out if the patch here doesn't work. 

What I'm saying is that deleting pods might not be a valid way of testing if `--previous` works. It will work in the case that the pod has crashed and we want to get the logs from when the pod containing Cilium crashed previously. Deleting pods and re-adding them in the Chaos tests is fine; we can get logs manually within the test in that case if needed. For the other times in which we do not delete pods, we can get the logs with this patch. Thus, I don't see the issue here.

Reply:
I mean, I tried a few times and I never get the logs, deleted, applying a new container, updating the image, always `Error from server (BadRequest): previous terminated container "cilium-agent" in pod...` You added the logs for one case (Cilium crash), but it doesn't work for the cases where cilium do not crash, but test fails, at that moment you will not have the cilium logs.

Reply:
@eloycoto @ianvernon The code is the following:
```golang
	deadlockCheckCmd = fmt.Sprintf("%s -n %s --timestamps=true --previous logs -l k8s-app=cilium | grep -qi -B 5 -A 5 deadlock", KubectlCmd, KubeSystemNamespace)
	res = kub.Exec(deadlockCheckCmd)
	if res.WasSuccessful() {
		log.Errorf("Deadlock during test run detected, check Cilium logs for context")
	}
```

even if the pod does not exist the message `"Deadlock during test run detected, check Cilium logs for context"` is not printed.

Also. I've noticed this messages are going to be logged. Is there anywhere where the logs are being grep for the "Deadlock" word?

Reply:
> even if the pod does not exist the message "Deadlock during test run detected, check Cilium logs for context" is not printed.

Fine by me, this is the correct behavior.

>Also. I've noticed this messages are going to be logged. Is there anywhere where the logs are being grep for the "Deadlock" word? 

Right now, there is no failure if 'Deadlock' is found in the logs. We could move this function to the `AfterEach` for the tests and check regardless of whether the test failed, and then fail if 'Deadlock' is found in the logs.

Reply:
> You added the logs for one case (Cilium crash), but it doesn't work for the cases where cilium do not crash, but test fails, at that moment you will not have the cilium logs.

This PR and what you are talking about are separate cases. I don't see the problem here; this can be tackled separately.

Reply:
Ok, I understand now, sorry for bothering!  Good for me